Feat(examples): XCC example of calling SocialDB from Aurora

birchmd opened the pull request:
In this PR we make the `aurora-solidity-sdk` functional, include an example of how to use it and add documentation describing how the Aurora Engine's XCC feature works.

This is once again a large PR, but for the benefit of the reviewers please focus on:

* `docs/NearFromAurora.md` (Is it readable? Does it answer the questions you have about XCC?)
* `examples/social-db-from-aurora` (This is code an external developer would need to write to use XCC; is it reasonable? Assume they have an understanding of both Near and EVM)

The code in `aurora-solidity-sdk` is lifted from the native ERC-20 connector, so reviewing that is less important (it has been reviewed once already). In the future we will need to change the native ERC-20 connector to depend on this repo instead of duplicating the SDK.

Casuso commented on examples/social-db-from-aurora/contract/src/SocialDB.sol:
```diff
@@ -0,0 +1,55 @@
+        wNEAR.transferFrom(msg.sender, address(this), attachedNear);
+
+        PromiseCreateArgs memory callSet =
+            near.call(socialdbAccountId, "set", data, attachedNear, SET_NEAR_GAS);
```
The sol contract looks clean and simple! Adding a couple of lines in the README file about the `set` method on `social-db` would be nice so devs can quickly understand what the purpose of the `set` method here is.

birchmd replied:
Done in https://github.com/aurora-is-near/aurora-contracts-sdk/pull/4/commits/33dcfbc3f27ea2e54b8070cc7e82a617244f0c1b

Casuso commented on docs/NearFromAurora.md:
```diff
@@ -0,0 +1,141 @@
+# Calling Near from Aurora, a.k.a Cross-Contract Calls (XCC)
```
The explanation is clear but the general process is complex. Maybe adding a diagram of the flow can help them significantly.

Casuso reviewed (approved):
LGTM!

mooori reviewed (approved):
LGTM and great devX for Aurora->Near XCC! Leaving some comments where I think additional documentation could be helpful plus a few nits.

birchmd commented on aurora-solidity-sdk/src/Types.sol:
```diff
@@ -0,0 +1,66 @@
+}
+
+enum PromiseArgsVariant
+/// Basic NEAR promise
```
`forge fmt` (the formatter that comes with the `forge` tool) puts the comments in this position. I agree it looks strange, but I will leave it this way since I don't want to have to remember to correct this every time after running the formatter.

birchmd commented on aurora-solidity-sdk/src/Types.sol:
```diff
@@ -0,0 +1,66 @@
+    Create,
+    /// NEAR promise with a callback attached.
+    Callback,
+    /// Description of arbitrary NEAR promise. Allows applying combinators
+    /// recursively, multiple action types and batched actions.
+    Recursive
```
Yes, that's right. But I describe it as recursive because the combinators can be used to create complex call trees (`then` pawns children, `and` joins multiple concurrent operations). The way this is modelled internally is with a self-referential type (like how you might model other graphs/trees).

birchmd commented on docs/NearFromAurora.md:
```diff
@@ -0,0 +1,141 @@
+This is to respect the revert semantics of the EVM; a revert should undo any effects caused by a call and any sub-calls it made.
+If the XCC precompile were to eagerly cause NEAR calls to happen then there would be no way to cancel them in the case of a revert.
+This is because NEAR cross-contract calls happen asynchronously.
+While each synchronous contract execution is atomic, the whole asynchronous transaction is not; if an asynchronous call fails it does not revert the effects of the prior calls that made it.
```
Added link to that page in https://github.com/aurora-is-near/aurora-contracts-sdk/pull/4/commits/33dcfbc3f27ea2e54b8070cc7e82a617244f0c1b

birchmd commented on docs/NearFromAurora.md:
```diff
@@ -0,0 +1,141 @@
+The `call` method can be used by any NEAR contract to interact with the EVM, not just the special sub-accounts created by XCC.
+The mechanism the `call` method uses to assign an EVM address to the predecessor account ID is the same for all accounts (no special case is made for XCC sub-accounts).
+This leads to the quirk we mentioned above that the sender address in the callback is not equal to the address that initiated the XCC.
+The way the address is calculated is by taking the last 20 bytes of the keccak hash of the NEAR account ID (see [Aurora contract code](https://github.com/aurora-is-near/aurora-engine/blob/2.8.1/engine-sdk/src/types.rs#L25-L27)).
```
Added short explanation in https://github.com/aurora-is-near/aurora-contracts-sdk/pull/4/commits/33dcfbc3f27ea2e54b8070cc7e82a617244f0c1b

birchmd commented on docs/NearFromAurora.md:
```diff
@@ -0,0 +1,141 @@
+Next, in both `Eager` and `Delayed` cases, there is the data which describes the NEAR call to be performed, `PromiseArgs`.
+This is an enum with variants that themselves are enums, but essentially there are: calls to a single NEAR contract (`PromiseCreateArgs`), a call with single callback which are useful to make a single NEAR call and then callback to Aurora (`PromiseWithCallbackArgs`), and a recursive data type (`NearPromise`) that captures arbitrary promise combinators (multiple callbacks, sums of multiple calls,  ...
+For most applications the `PromiseWithCallbackArgs` should be sufficient.
+If your application needs the more general `NearPromise` then see the [NEAR documentation](https://nomicon.io/RuntimeSpec/Components/BindingsSpec/PromisesAPI) for more information.
```
Reorganized as doc comments in https://github.com/aurora-is-near/aurora-contracts-sdk/pull/4/commits/33dcfbc3f27ea2e54b8070cc7e82a617244f0c1b

birchmd commented on examples/social-db-from-aurora/integration-tests/src/lib.rs:
```diff
@@ -0,0 +1,230 @@
+            .unwrap()
+            .into_result()
+            .unwrap();
+        social_db
+            .call("set_status")
+            .args_json(serde_json::json!({"status": "Live"}))
+            .transact()
+            .await
+            .unwrap()
+            .into_result()
+            .unwrap();
```
Done in https://github.com/aurora-is-near/aurora-contracts-sdk/pull/4/commits/33dcfbc3f27ea2e54b8070cc7e82a617244f0c1b

birchmd commented on examples/social-db-from-aurora/contract/src/SocialDB.sol:
```diff
@@ -0,0 +1,55 @@
+// SPDX-License-Identifier: CC-BY-1.0
```
Done in https://github.com/aurora-is-near/aurora-contracts-sdk/pull/4/commits/33dcfbc3f27ea2e54b8070cc7e82a617244f0c1b

birchmd commented on examples/social-db-from-aurora/integration-tests/src/lib.rs:
```diff
@@ -0,0 +1,230 @@
+                user_account.id().as_bytes(),
+            );
+
+        // Deploy EVM contract that will be the proxy on Aurora for the Social DB contract on Near
```
Added in https://github.com/aurora-is-near/aurora-contracts-sdk/pull/4/commits/33dcfbc3f27ea2e54b8070cc7e82a617244f0c1b

birchmd commented on aurora-solidity-sdk/src/AuroraSdk.sol:
```diff
@@ -0,0 +1,217 @@
+// Address of predecessor account id precompile in Aurora.
+// It allows getting the predecessor account id of the current call.
+address constant PREDECESSOR_ACCOUNT_ID_PRECOMPILE = 0x723FfBAbA940e75E7BF5F6d61dCbf8d9a4De0fD7;
+// Address of predecessor account id precompile in Aurora.
```
Fixed in https://github.com/aurora-is-near/aurora-contracts-sdk/pull/4/commits/33dcfbc3f27ea2e54b8070cc7e82a617244f0c1b

birchmd commented on aurora-solidity-sdk/src/AuroraSdk.sol:
```diff
@@ -0,0 +1,217 @@
+        return string(returnData);
+    }
+
+    /// Crease a base promise. This is not immediately schedule for execution
```
Fixed in https://github.com/aurora-is-near/aurora-contracts-sdk/pull/4/commits/33dcfbc3f27ea2e54b8070cc7e82a617244f0c1b

birchmd commented on aurora-solidity-sdk/src/AuroraSdk.sol:
```diff
@@ -0,0 +1,217 @@
+    }
+
+    /// Similar to `call`. It is a wrapper that simplifies the creation of a promise
+    /// to a controct inside `Aurora`.
```
Fixed in https://github.com/aurora-is-near/aurora-contracts-sdk/pull/4/commits/33dcfbc3f27ea2e54b8070cc7e82a617244f0c1b

birchmd commented on aurora-solidity-sdk/src/AuroraSdk.sol:
```diff
@@ -0,0 +1,217 @@
+    }
+
+    /// Schedule a base promise to be executed on NEAR. After this function is called
+    /// the promise should not be used anymore.
```
It cannot be used to obtain the result because the result is always only obtainable from the `promiseResult` function inside a callback. I've added a brief mention of this to the comment in https://github.com/aurora-is-near/aurora-contracts-sdk/pull/4/commits/33dcfbc3f27ea2e54b8070cc7e82a617244f0c1b

birchmd commented:
Thanks for the reviews @mooori and @Casuso ! I think I've addressed all the comments, so I'll merge this PR for now. I'm sure we'll iterate on this documentation more as it gets usage from real developers.
